### PR TITLE
Add What's Here to Array RDoc

### PR DESCRIPTION
**Pull request by BurdetteLamar**

### CODE REVIEW

File: array.c [hunk 1]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  #bsearch:: Returns an element selected via a binary search
+ *             as determined by a given block.
+ *  #bsearch_index:: Returns the index of an element selected via a binary search
+ *                   as determined by a given block.
+ *  #sample:: Returns one or more random elements.
+ *  #shuffle:: Returns elements in a random order.
+ *
+ *  === Methods for Assigning
+ *
+ *  These methods add or replace elements in +self+.
```
Comment (jeremyevans) [inline array.c:8160]: I would change to `add, replace, or reorder`, as there are many methods listed that reorder and do not add or replace.

File: array.c [hunk 2]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  #concat:: Appends all elements from given arrays.
+ *  #fill:: Replaces specified elements with specified objects.
+ *  #replace:: Replaces the content of +self+ with the content of a given array.
+ *  #reverse!:: Replaces +self+ with its elements reversed.
+ *  #rotate!:: Replaces +self+ with its elements rotated.
+ *  #shuffle!:: Replaces +self+ with its elements in random order.
+ *  #sort!:: Replaces +self+ with its elements sorted,
+ *           as determined by <tt><=></tt> or a given block.
+ *  #sort_by!:: Replaces +self+ with its elements sorted, as determined by a given block.
+ *  #slice!:: Removes and returns a sequence of elements.
```
Comment (jeremyevans) [inline array.c:8175]: Since `slice!` deletes, it should probably be moved to the `Methods for Deleting` section.

File: array.c [hunk 3]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *
+ *  === Methods for Deleting
+ *
+ *  Each of these methods removes elements from +self+:
```
Comment (jeremyevans) [inline array.c:8179]: `compact!` should probably be listed in this section.

File: array.c [hunk 4]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  #delete:: Removes elements equal to a given object.
+ *  #delete_at:: Removes the element at a given offset.
+ *  #delete_if:: Removes elements specified by a given block.
+ *  #keep_if:: Removes elements not specified by a given block.
+ *  #reject!:: Removes elements specified by a given block.
+ *  #select!:: Removes elements not specified by a given block.
+ *  #uniq!:: Removes duplicates.
+ *  #filter!:: Removes elements not specified by a given block.
+ *
+ *  Each of these methods returns a copy of +self+ with some elements removed.
```
Comment (jeremyevans) [inline array.c:8192]: Arguably, each of these could be listed under `Methods for Fetching`.  However, if we want to keep this section, we should probably move `slice` to it, for consistency with other methods listed.

File: array.c [hunk 5]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  Each of these methods removes elements from +self+:
+ *
+ *  #pop:: Removes and returns the last element.
+ *  #shift::  Removes and returns the first element.
+ *  #delete:: Removes elements equal to a given object.
+ *  #delete_at:: Removes the element at a given offset.
+ *  #delete_if:: Removes elements specified by a given block.
+ *  #keep_if:: Removes elements not specified by a given block.
+ *  #reject!:: Removes elements specified by a given block.
+ *  #select!:: Removes elements not specified by a given block.
```
Comment (jeremyevans) [inline array.c:8188]: `select!` and `filter!` are aliases and should be listed on the same line.

File: array.c [hunk 6]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  === Other Methods
+ *
+ *  #*:: Returns one of the following:
+ *       - With integer argument +n+, a new array that is the concatenation
+ *         of +n+ copies of +self+.
+ *       - With string argument +field_separator+, a new string that is equivalent to
+ *         <tt>join(field_separator)</tt>.
+ *  #abbrev:: Returns a hash of unambiguous abbreviations for elements.
+ *  #map, #collect:: Returns an array containing the block return-value for each element.
+ *  #map!, #collect!:: Replaces each element with a block return-value.
```
Comment (jeremyevans) [inline array.c:8251]: I would consider moving the map/collect methods to the `Methods for Converting` section.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Thanks for working on this. Most of these changes look good, but I have some changes requested.

**Comment by BurdetteLamar** [comment]
Thanks, @jeremyevans, for this careful review.

File: array.c [hunk 7]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
  *     arr = [1, 2, 3, 4, 5, 6]
  *     arr.keep_if {|a| a < 4}   #=> [1, 2, 3]
  *     arr                       #=> [1, 2, 3]
+ *
+ *  == What's Here
+ *
+ *  \Class \Array provides methods that are useful for:
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c:8080]: Class, as used here, doesn't refer to the class Class, and so should not link?
Also, I thought we were agreeing that things in class Array should not link to Array?

File: array.c [hunk 8]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  - {Deleting}[#class-Array-label-Methods+for+Deleting]
+ *  - {Combining}[#class-Array-label-Methods+for+Combining]
+ *  - {Iterating}[#class-Array-label-Methods+for+Iterating]
+ *  - {Converting}[#class-Array-label-Methods+for+Converting]
+ *  - {And more....}[#class-Array-label-Other+Methods]
+ *
+ *  === Methods for Creating an \Array
+ *
+ *  ::[]:: Returns a new array populated with given objects.
+ *  ::new:: Returns a new empty array.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c [hunk 9]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *          element in a given object.
+
+ *  === Methods for Fetching
+ *
+ *  These methods do not modify +self+.
+ *
+ *  #[]:: Returns one or more elements.
+ *  #fetch:: Returns the element at a given offset.
+ *  #first:: Returns one or more leading elements.
+ *  #last:: Returns one or more trailing element.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c [hunk 10]
```diff
@@ -8074,6 +8074,183 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  #max:: Returns one or more maximum-valued elements,
+ *         as determined by <tt><=></tt> or a given block.
+ *  #max:: Returns one or more minimum-valued elements,
+ *         as determined by <tt><=></tt> or a given block.
+ *  #minmax:: Returns the minimum-valued and maximum-valued elements,
+ *            as determined by <tt><=></tt> or a given block.
+ *  #assoc:: Returns the first element that is an array
+ *           whose first element <tt>==</tt> a given object.
+ *  #assoc:: Returns the last element that is an array
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c]: Revised.

File: array.c (hunk 1 above)
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c (hunk 2 above)
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c (hunk 3 above)
Comment (BurdetteLamar) [reply array.c:8181]: Fixed.

File: array.c (hunk 4 above)
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c (hunk 5 above)
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c (hunk 6 above)
Comment (BurdetteLamar) [reply array.c:8248]: Moved.

File: array.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c]: Nevertheless, I've removed the three suppressions here in What's Here.

File: array.c [hunk 11]
```diff
@@ -8074,6 +8074,180 @@ rb_ary_deconstruct(VALUE ary)
...
+ *
+ *  === Other Methods
+ *
+ *  #*:: Returns one of the following:
+ *       - With integer argument +n+, a new array that is the concatenation
+ *         of +n+ copies of +self+.
+ *       - With string argument +field_separator+, a new string that is equivalent to
+ *         <tt>join(field_separator)</tt>.
+ *  #abbrev:: Returns a hash of unambiguous abbreviations for elements.
+ *  #map!, #collect!:: Replaces each element with a block return-value.
```
Comment (jeremyevans) [inline array.c:8248]: Sorry for not being completely clear, but `map!`/`collect!` should probably also be moved to the `Methods for Converting` section.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks good, only a couple small changes requested, then it is OK to merge.

File: array.c [hunk 12]
```diff
@@ -8074,6 +8074,180 @@ rb_ary_deconstruct(VALUE ary)
...
+ *
+ *  #pop:: Removes and returns the last element.
+ *  #shift::  Removes and returns the first element.
+ *  #compact!:: Removes all non-+nil+ elements.
+ *  #delete:: Removes elements equal to a given object.
+ *  #delete_at:: Removes the element at a given offset.
+ *  #delete_if:: Removes elements specified by a given block.
+ *  #keep_if:: Removes elements not specified by a given block.
+ *  #reject!:: Removes elements specified by a given block.
+ *  #select!, #filter:: Removes elements not specified by a given block.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c]: Oh, wow.  That *is* careless.  Now fixed.

File: array.c (hunk 11 above)
Comment (BurdetteLamar) [reply array.c]: Done.

**Review by jeremyevans — APPROVED** [review]

**Pull request merged**